configure

```diff
@@ -0,0 +1,360 @@
+#!/bin/sh
+### Custom configure script of Dinit.
+
+## Initial prepartion
+cd "$(dirname "$0")" || echo "Can't cd to $(dirname "$0")"
+. build/version.conf
+
+## Helper funcations
+# I suggest do not comment this thing but if your system
+# is don't have `printf` you need to comment this thing.
+# See http://www.etalabs.net/sh_tricks.html
+echo() {
+    IFS=" "
+    printf %s\\n "$*"
+}
+
+info() {
+    echo "Info: $*"
+}
+
+sub_info() {
+    echo "  ... $*"
+}
+
+error() {
+    echo "Error: $*"
+    exit 1
+}
+
+sub_error() {
+    echo "  ... Error: $*"
+    exit 1
+}
+
+warning() {
+    echo
+    echo "Warning: $*"
+}
+
+sub_warning() {
+    echo "  ... $*"
+    echo
+}
+
+strip_dir() {
+    cd "$2" || error "Can't cd to $2"
+    eval "$1=\"\$PWD\""
+    eval "$1=\${$1# }"
+    cd "$BASE_DIR" || error "Can't cd to $BASE_DIR"
+}
+
+create_testfile_cc() {
+    echo "int main(int argc, char **argv) { return 0; }" > testfile.cc
+}
+
+findcxx() {
+    type "$1" > /dev/null 2>&1
+    if test $? = 0; then
+        if [ "$2" = "host" ]; then
+            hostcompiler="$1"
+        else
+            compiler="$1"
+        fi
+    fi
+}
+
+cxx_works() {
+    info Checking C++ compiler works...
+    $compiler -o testfile.o testfile.cc
+    if [ $? != 0 ]; then
+        rm -f testfile*
+        sub_error Seems like C++ compiler is not working!
+    else
+        rm -f testfile
+        sub_info Yes.
+    fi
+}
+
+find_program() {
+    if [ -x $(command -v "$1") ]; then
+        sub_info Yes.
+    else
+        sub_error "$1" is not executable/exist!
+    fi
+}
+
+try_cxx_argument() {
+    info Checking whatever compiler accpets $2...
+    if $compiler -c $CPPFLAGS $CXXOPTS $2 $CUSTOM_CXXOPTS testfile.cc -o testfile.o > /dev/null 2>&1 ; then
+        rm testfile.o
+        sub_info Yes.
+        eval "$1=\"\${$1} \$2\""
+        eval "$1=\${$1# }"
+    else
+        if [ "$2" = "-std=c++11" ]; then
+            sub_error "Seems like C++ compiler don't support $2 but It's required!"
+        else
+            sub_info No.
+        fi
+    fi
+}
+
+try_ld_argument() {
+    info Checking whatever linker accepts $2...
+    if $compiler $CPPFLAGS $CXXOPTS $CUSTOM_CXXOPTS $2 testfile.cc -o testfile.o > /dev/null 2>&1 ; then
+        sub_info Yes.
+        eval "$1=\"\${$1} \$2\""
+        eval "$1=\${$1# }"
+    else
+        sub_info No.
+    fi
+}
+
+usage() {
+    cat << EOF
+Usage: $0 [OPTION]...
+
+Defaults for the options are specified in brackets.
+
+  --help                        This help message.
+  --quiet                       Don't print normal messages, just errors.
+  --clean                       Clear mconfig and configure's temp files.
+
+Target options:
+  --target                      Set the target system (for cross-compiling) []
```

Comment:
It's from `autotools` (specially `autoconf`) legacy. I writed this for hold compatibility with `autoconf` generated `configure`s.
See: https://www.gnu.org/software/automake/manual/html_node/Cross_002dCompilation.html
Use it is simple:
```
$ configure --target=aarch64-linux-gnu --build=x86_64-linux-gnu
```
or
```
$ configure --host=i686-linxu-gnu --build=x86_64-linux-gnu
```
or
```
$ configure --target=aarch64-linxu-gnu
```

Reply:
For autotools it is potentially used for more than just selecting the compiler/build tools chain. But anyway I don't think we should be concerned about autotools compatibility - we're not setting out to remake autotools.

For example, it doesn't make sense that there is both `--target` and `--platform`. What if I specify:

`--target=i686-linux-gnu --platform=FreeBSD`?

Currently it is possible to specify compiler, target, and platform - but target is meaningless if compiler is specified, so I think it's better not to have it. Otherwise it's not clear if I specify `--target` do I also need to specify `--platform` for example? (I mean: I understand what they do. But for other users, it will not be clear).

Reply:
`--platform` just for don't set some arguments to compiler/linker. So "For cross-compiling just set CXX and HOSTCXX options" right?

Reply:
Well selecting the platform also selects suitable defaults for some of the other options. So I would say for cross compiling you should specify CXX, HOSTCXX, and `--platform`.

Reply:
autotools specifically describes --target as only useful or meaningful when compiling a compiler.

You build on the --build OS, you cross compile to produce binaries that run on the --host OS, and when you copy those built binaries over to the host OS and run them there, they then output machine code for the --target OS.

In nearly all cases you only care about the build and host OSes.

Reply:
BTW: $HOSTCXX is for the --build machine, and autotools would call it $CXX_FOR_BUILD.

Reply:
Yeah this is true, autotools has very specific host/build/target meaning that we are conflicting with here. (That's my bad, I think of it in terms of "build native code for the machine that _hosts_ the build", but that's not the autotools usage). It would be better not to re-use the same names as autotools but in a different way; that could cause confusion.

I don't like "CXX_FOR_BUILD" much since it's ambiguous. But it's hard to come up with a short name that isn't. So I guess we may as well go with this.

@mobin-2008 let's rename HOSTCXX etc to CXX_FOR_BUILD - I'll do the same in existing mconfig files / makefiles, if you make the same changes here. So it will be:

CXX_FOR_BUILD
CXXFLAGS_FOR_BUILD
CPPFLAGS_FOR_BUILD
LDFLAGS_FOR_BUILD

I notice the help output only has HOSTCXX ( = CXX_FOR_BUILD), you should probably add the others also.